### PR TITLE
feat: Introduce workflow orchestration

### PR DESCRIPTION
**Pull request by adrien2p**
Merged with https://github.com/medusajs/medusa/pull/6128

### CODE REVIEW

**Comment by carlos-r-l-rodrigues** [comment]
note: More tests will be introduced to the module in a follow up PR integrating it with `/medusa`

File: packages/modules-sdk/src/definitions.ts [hunk 1]
```diff
@@ -15,6 +15,7 @@ export enum Modules {
   PRICING = "pricingService",
   PROMOTION = "promotion",
   AUTHENTICATION = "authentication",
+  WORKFLOW_ORCHESTRATOR = "workflowOrchestrator",
```
Comment (olivermrbl) [inline packages/modules-sdk/src/definitions.ts]: thought(non-blocking): Think we should find a simpler module name for when we release the package publicly. Not important to act on right now.

File: packages/types/src/workflow-orchestrator/common/workflow-execution.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { BaseFilterable } from "../../dal"
```
Comment (olivermrbl) [inline packages/types/src/workflow-orchestrator/common/workflow-execution.ts]: suggestion: Create `mutations.ts` file for `UpsertWorkflowExecutionDTO` to follow recently established convention

File: packages/workflows-sdk/src/utils/composer/create-step.ts [hunk 3]
```diff
@@ -264,7 +279,8 @@ export function createStep<
   compensateFn?: CompensateFn<TInvokeResultCompensateInput>
 ): StepFunction<TInvokeInput, TInvokeResultOutput> {
   const stepName =
-    (isString(nameOrConfig) ? nameOrConfig : nameOrConfig.name) ?? invokeFn.name
+    (isString(nameOrConfig) ? nameOrConfig : nameOrConfig.action) ??
```
Comment (olivermrbl) [inline packages/workflows-sdk/src/utils/composer/create-step.ts]: question: Should we call `action` something tighter coupled with what it is e.g. `stepName` or `name` or is there a specific reason for the change?

Comment (olivermrbl) [reply packages/workflows-sdk/src/utils/composer/create-step.ts]: I personally like `name` the most to be consistent with the `nameOrConfig` and general `createStep` method signature

File: packages/workflow-orchestrator-inmemory/src/scripts/seed.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
+import { Modules } from "@medusajs/modules-sdk"
```
Comment (olivermrbl) [inline packages/workflow-orchestrator-inmemory/src/scripts/seed.ts]: suggestion: Use recently introduced utils

File: packages/workflow-orchestrator-inmemory/src/scripts/migration-up.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
+import { Modules } from "@medusajs/modules-sdk"
```
Comment (olivermrbl) [inline packages/workflow-orchestrator-inmemory/src/scripts/migration-up.ts]: suggestion: Use recently introduced utils

File: packages/workflow-orchestrator-inmemory/src/scripts/migration-down.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import { Modules } from "@medusajs/modules-sdk"
```
Comment (olivermrbl) [inline packages/workflow-orchestrator-inmemory/src/scripts/migration-down.ts]: suggestion: Use recently introduced utils

File: packages/workflow-orchestrator-inmemory/src/models/workflow-execution.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  Unique,
+} from "@mikro-orm/core"
+
+type OptionalFields = "deleted_at"
+
+@Entity()
+@Unique({
+  name: "IDX_workflow_execution_workflow_id_transaction_id_unique",
+  properties: ["workflow_id", "transaction_id"],
+})
```
Comment (olivermrbl) [inline packages/workflow-orchestrator-inmemory/src/models/workflow-execution.ts]: question: Do we need to add?
```
@Filter(DALUtils.mikroOrmSoftDeletableFilterOptions)
```

File: packages/workflow-orchestrator-inmemory/package.json [hunk 8]
```diff
@@ -0,0 +1,64 @@
+{
+  "name": "@medusajs/workflow-orchestrator-inmemory",
```
Comment (olivermrbl) [inline packages/workflow-orchestrator-inmemory/package.json]: Same as other comment about name change. Not important right now.

**Review by olivermrbl — APPROVED** [review]

File: packages/workflows-sdk/src/utils/composer/create-step.ts (hunk 3 above)
Comment (carlos-r-l-rodrigues) [reply packages/workflows-sdk/src/utils/composer/create-step.ts]: Will normalize it to `name`. I just changed to use the same property as in the transaction orchestrator definition. But it makes sense to have more meaninful names for the Workflows.

File: packages/modules-sdk/src/definitions.ts (hunk 1 above)
Comment (carlos-r-l-rodrigues) [reply packages/modules-sdk/src/definitions.ts]: Any suggestion? just `workflows` or something like that?

Comment (olivermrbl) [reply packages/modules-sdk/src/definitions.ts]: Yeah, let's go with that for now 👍

**Comment by carlos-r-l-rodrigues** [comment]
@olivermrbl @adrien2p, I've renamed this module. Do you guys mind to have a quick look if it is alright to merge it like that?

**Comment by adrien2p** [comment]
> @olivermrbl @adrien2p, I've renamed this module. Do you guys mind to have a quick look if it is alright to merge it like that?

Since this is suppose to be the main entry to run any kind of workflow, should we test something like workflow runner (by default in memory) and then we can have workflow runner redis etc? Wdyt?

**Pull request closed without merging**